ComponentKit/Utilities/CKComponentAction.h

```diff
@@ -11,44 +11,180 @@
 #import <UIKit/UIKit.h>
 
 #import <ComponentKit/CKComponentViewAttribute.h>
+#import <ComponentKit/CKComponentActionInternal.h>
 
 @class CKComponent;
 
 /**
- A component action is simply a selector. The selector may optionally take one argument: the sending component.
+ CKTypedComponentAction is a struct that represents a method invocation that can be passed to a child component to
+ trigger a method invocation on a target.
+ 
+ We allow a typed specification of the parameters that will be provided as arguments to the component action at runtime
+ through the variadic templated arguments. You may specify an arbitrary number of arguments to your component action,
+ and you may use either object, or primitive arguments. Only trivially-constructible arguments or pointers can be used
+ as types for component actions, so primitive types like int, CGRect, and NSObject * are fine. This is enforced via a
+ compile time check.
+
+ Methods will always be provided the sender as the first argument.
+ 
+ Usage in your component header:
+ 
+ @interface MyComponent : CKComponent
+ + (instancetype)newWithAction:(CKTypedComponentAction<NSString *, int>)action;
+ @end
+ 
+ When creating the action:
+ 
+ Option 1 - Scope action. Similar to target/selector action in that it skips the responder chain from the sender, and
+            directly invokes the selector on the component or controller corresponding with the scope. Promotion is also
+            supported for scope-based actions. Scope actions weakly capture the component or controller. Does not
+            depend on the mount-based responder chain to call on the component or controller. Action may be implemented
+            on either component or controller.
+
+             + (instancetype)new
+             {
+               CKComponentScope scope(self);
+               return [super
+                       newWithComponent:
+                        [MyComponent
+                         newWithAction:{scope, @selector(methodWithSender:)}]];
+             }
+             - (void)methodWithSender:(CKComponent *)sender {}
+ 
+ Option 2 - Target/selector action. Ensures that the target responds to the given selector. Target must directly
```

Comment:
This might be needed, but I think the first option seems the best and the pattern that should be encouraged the most - it tends to ensure that actions are handled in the component hierarchy whenever possible as well. Would we want to discourage Option 2, or at least suggest the usage of Option 1 over this one?